std/range/primitives.d

```diff
@@ -1385,29 +1385,30 @@ template hasLvalueElements(R)
 }
 
 /**
-Returns $(D true) if $(D R) has a $(D length) member that returns an
-integral type. $(D R) does not have to be a range. Note that $(D
-length) is an optional primitive as no range must implement it. Some
-ranges do not store their length explicitly, some cannot compute it
-without actually exhausting the range (e.g. socket streams), and some
-other ranges may be infinite.
-
-Although narrow string types ($(D char[]), $(D wchar[]), and their
-qualified derivatives) do define a $(D length) property, $(D
-hasLength) yields $(D false) for them. This is because a narrow
-string's length does not reflect the number of characters, but instead
-the number of encoding units, and as such is not useful with
-range-oriented algorithms.
- */
-template hasLength(R)
-{
-    enum bool hasLength = !isNarrowString!R && is(typeof(
+Yields `true` if `R` has a `length` member that returns a value of either
+`size_t` or `ulong` type. `R` does not have to be a range. If `R` is a range,
+algorithms in the standard library are only guaranteed to support `length` with
+type `size_t`.
+
+Note that `length` is an optional primitive as no range must implement it. Some
+ranges do not store their length explicitly, some cannot compute it without
+actually exhausting the range (e.g. socket streams), and some other ranges may
+be infinite.
+
+Although narrow string types (`char[]`, `wchar[]`, and their qualified
+derivatives) do define a `length` property, `hasLength` yields `false` for them.
+This is because a narrow string's length does not reflect the number of
+characters, but instead the number of encoding units, and as such is not useful
+with range-oriented algorithms. To use strings as random-access ranges with
+length, use $(XREF string, representation) or $(XREF utf, byCodeUnit). */
```

Comment:
The `XREF` macro was deprecated, please use the universal `REF`, e.g. `$(REF, byCodeUnit, std, utf)